src/Microsoft.AspNetCore.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -167,8 +169,10 @@ public MemoryPoolIterator ConsumingStart()
             {
                 throw new InvalidOperationException("Already consuming input.");
             }
-
-            return new MemoryPoolIterator(_head);
+            lock (_sync)
```

Comment:
Update `_consumingState` inside the `_sync` lock here and in `ConsumingComplete` to avoid a race in dispose causing blocks to not get returned.